assembly/src/parsers/ast/mod.rs

```diff
@@ -151,104 +201,111 @@ impl Deserializable for ProcedureAst {
 
 /// Parses the provided source into a program AST. A program consist of a body and a set of
 /// internal (i.e., not exported) procedures.
-#[cfg(test)]
-pub fn parse_program(source: &str) -> Result<ProgramAst, AssemblyError> {
-    let mut tokens = TokenStream::new(source)?;
-    let imports = parse_imports(&mut tokens)?;
-
-    let mut context = ParserContext {
-        imports,
-        ..Default::default()
-    };
-
-    context.parse_procedures(&mut tokens, false)?;
-
-    // make sure program body is present
-    let next_token = tokens
-        .read()
-        .ok_or_else(|| AssemblyError::unexpected_eof(tokens.pos()))?;
-    if next_token.parts()[0] != Token::BEGIN {
-        return Err(AssemblyError::unexpected_token(next_token, Token::BEGIN));
-    }
+impl TryFrom<&str> for ProgramAst {
```

Comment:
I like this approach for parsing, but I have a couple of thoughts/questions:

1. This should probably be a part of a different (smaller) PR as this PR is going to be complex enough to review and this is a pretty self-contained.
2. I wonder if this approach will be flexible enough. Specifically, we might want to introduce some parsing options (e.g., whether to retain comments or not). I am not sure if we'll do it, but if we do, the `TryFrom` approach would probably not be able to accomodate this.

Reply:
Change reverted